src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java

```diff
@@ -245,6 +280,10 @@ private static void applyPatchToFile(
       oldContent = new ArrayList<>();
     } else {
       oldContent = readFile(inputFile);
+      // Preserve old file permission if no explicit permission is set.
+      if (filePermission == -1) {
```

Comment:
If you want to use some more exotic Java data types, you could use `Optional<BitSet>` for the filePermission type (and `BitSet` in general, rather than an `int`, to store the permissions). Then you could use an `enum` as the keys for the `BitSet` (e.g. `EXECUTABLE = 0, WRITABLE = 1, READABLE = 2`).

But using `-1` as the indicator for "no explicit permissions" is also fine, if you prefer it. ;)

Reply:
I tried to use BitSet, but found converting a character (eg. `7`) to a BitSet also requires some steps. The code overall didn't look much cleaner, so I'll keep use `-1` and bit operations ;)